L/LibGit2/bundled/patches/libgit2-hostkey.patch

```diff
@@ -1,48 +1,16 @@
-diff --git a/include/git2/cert.h b/include/git2/cert.h
-index e8cd2d180..54293cd31 100644
---- a/include/git2/cert.h
-+++ b/include/git2/cert.h
-@@ -111,6 +111,14 @@ typedef struct {
- 	 * have the SHA-256 hash of the hostkey.
- 	 */
- 	unsigned char hash_sha256[32];
-+
-+	/**
-+	 * Hostkey itself.
-+	 */
-+	int hostkey_type;
-+	size_t hostkey_len;
-+	unsigned char hostkey[1024];
-+
- } git_cert_hostkey;
- 
- /**
 diff --git a/src/transports/ssh.c b/src/transports/ssh.c
-index f4ed05bb1..ec6366a5f 100644
+index 471c3273ed..32189d0979 100644
 --- a/src/transports/ssh.c
 +++ b/src/transports/ssh.c
-@@ -523,6 +523,7 @@ static int _git_ssh_setup_conn(
+@@ -525,6 +525,7 @@ static int _git_ssh_setup_conn(
  	git_credential *cred = NULL;
- 	LIBSSH2_SESSION* session=NULL;
- 	LIBSSH2_CHANNEL* channel=NULL;
+ 	LIBSSH2_SESSION *session=NULL;
+ 	LIBSSH2_CHANNEL *channel=NULL;
 +	char *host_and_port;
  
  	t->current_stream = NULL;
  
-@@ -566,6 +567,12 @@ post_extract:
- 
- 		cert.parent.cert_type = GIT_CERT_HOSTKEY_LIBSSH2;
- 
-+		key = libssh2_session_hostkey(session, &cert.hostkey_len, &cert.hostkey_type);
-+		bzero(&cert.hostkey, sizeof(cert.hostkey));
-+		if (cert.hostkey_len > sizeof(cert.hostkey))
-+			cert.hostkey_len = sizeof(cert.hostkey);
-+		memcpy(&cert.hostkey, key, cert.hostkey_len);
-+
- #ifdef LIBSSH2_HOSTKEY_HASH_SHA256
- 		key = libssh2_hostkey_hash(session, LIBSSH2_HOSTKEY_HASH_SHA256);
- 		if (key != NULL) {
```

Comment:
@StefanKarpinski This part of the patch seems to have been replaced by https://github.com/libgit2/libgit2/commit/29fe5f61ac2d5f7039ecf7dad834918fc683c98e. I'm not sure whether it's perfectly equivalent though. x-ref https://github.com/JuliaPackaging/Yggdrasil/pull/2419.

Reply:
Fun. I guess we need to make sure that the `certificate_callback` and `ssh_knownhost_check` stuff in `LibGit2` still works with the new libgit2 version. The patch was required, IIRC, because the old version of libgit2 was doing several broken things:

1. It was failing to pass the port number with the host, which is necessary when connecting over SSH using different ports and using the known hosts file.
2. It didn't pass the actual hostkey info from the connection to the callback, which made it impossible to do hostkey verification.

It's unclear how (or if) anyone was actually using SSH hostkey checking previously.

Reply:
It looks like the first thing is still broken but the second thing got fixed but in an incompatible way because they stuck that `raw_type` field in there instead of just appending the hostkey data.

Reply:
Thanks. So AFAIK I kept the patch for the first thing so it should be OK. For the second one, tests added by https://github.com/JuliaLang/julia/pull/39324 cover it, right? At least they pass with libgit2 1.3 and this branch.

Reply:
I don't think that tests the port part but it looks like it tests the hostkey part. I was testing by actually connecting to a host that I was proxying through to github.

Reply:
OK. I've just done this, and it seems to work!

For the record, here's what I did to forward port 22 on my machine to github.com:
```
sudo iptables -t nat -A PREROUTING -i wlo1 -p tcp --dport 32800  -j DNAT --to 38.140.156.5:22
sudo iptables -A FORWARD -i wlo1 -p tcp --dport 22 -d 38.140.156.5 -j ACCEPT
```
Then I had to go through my router as trying to connect from the same machine avoids the forwarding.

Then I installed a package via (after manually adding A.B.C.D to ~/.ssh/known_hosts with the signature for github.com):
```
]add ssh://git@A.B.C.D:32800/nalimilan/FreqTables.jl.git
```

One weird fact is that if I call `git clone ssh://git@A.B.C.D:32800/nalimilan/FreqTables.jl.git`, an entry for `[88.122.67.23]:32800` is added to ~/.ssh/known_hosts, but that entry isn't used by `]add ...` to work. I need to add the signature without the port. I guess that's expected since the same happens on Julia 1.6.

Anything I missed?

Reply:
Bump.